lib/plugin_manager.rb

```diff
@@ -0,0 +1,211 @@
+  end
+
+  def _sort_gemfile_plugins
+    # todo we have to take of the order..
+    # e.g. global roles first and
+    # reporting engine before costs/reporting
+  end
+
+  def _write_to_gemfile_plugins_file
+    File.open(GEMFILE_PLUGINS_PATH, 'w') do |f|
+      f.write @gemfile_plugins
+    end
+  end
+
+  def _bundle
+    Bundler.with_clean_env do
+      system 'bundle install --no-deployment'
+    end
+  end
+
+  def _migrate
+    system "RAILS_ENV='#{@environment}' bundle exec rake db:migrate"
+  end
+
+  def _revert_migrations(plugin)
+    migration_path = OpenProject::Application.config.paths['db/migrate'].select { |path| path.include?(plugin.name) }
+    ActiveRecord::Migrator.migrate migration_path, 0
+  end
+
+  def _assets_precompile
+    system "RAILS_ENV='production' bundle exec rake assets:precompile"
+  end
+
+  def _assets_webpack
+    system "RAILS_ENV='production' bundle exec rake assets:webpack"
+  end
+
+  def _assets_clobber
+    system "RAILS_ENV='production' bundle exec rake assets:clobber"
+  end
+
+  def _remove_plugin_from_gemfile_plugins_file(plugin)
+    _remove_plugin_from_gemfile_plugins(plugin)
+    _write_to_gemfile_plugins_file
+    _remove_gemfile_plugins_if_empty
+  end
+
+  def _remove_plugin_from_gemfile_plugins(plugin)
+    gemfile_plugins_new = ''
+    plugins = [plugin].concat _dependencies_only_required_by(plugin)
+    @gemfile_plugins.each_line do |line|
+      gemfile_plugins_new << line if _line_contains_no_plugin?(line, plugins)
+    end
+    @gemfile_plugins = gemfile_plugins_new
+  end
+
+  def _dependencies_only_required_by(plugin)
+    dependencies = plugin.dependencies
+    dependencies.select { |dependency| dependency._not_needed_by_any_other_than?(plugin) }
+  end
+
+  def _line_contains_no_plugin?(line, plugins)
+    result = true
+    plugins.each do |plugin|
+      result = false if plugin.included_in?(line)
+    end
+    result
+  end
+
+  def _remove_gemfile_plugins_if_empty
+    FileUtils.rm(GEMFILE_PLUGINS_PATH) if @gemfile_plugins == ''
+  end
+end
+
+class Plugin
+  def self._available?(name)
+    available_plugins[name]
+  end
+
+  def self.available_plugins
+    @available_plugins || _load_available_plugins
+  end
+
+  def self._load_available_plugins
+    unless File.exists?(PLUGINS_YML_PATH)
+      puts 'Could not find plugin list, abort!'
+      exit
+    end
+    YAML.load_file(PLUGINS_YML_PATH)
+  end
+
+  attr_reader :name
+
+  def initialize(name)
+    # This is necessary because we have some strange dependencies in openproject-pdf_export.
+    # When we delete those, we can delete this if.
+    if name == 'pdf-inspector'
+      @name = 'pdf-inspector'
+    else
+      unless Plugin._available?(name)
+        puts 'Could not find plugin, abort!'
+        exit
+      end
+      @name = name
+      @url = Plugin.available_plugins[name][:url]
+      @branch = Plugin.available_plugins[name][:branch]
+    end
+  end
+
+  def _not_needed_by_any_other_than?(plugin)
```

Comment:
Assignment Branch Condition size for _not_needed_by_any_other_than? is too high. [15.33/15]